tcmalloc/internal/percpu_rseq_aarch64.S

```diff
@@ -0,0 +1,487 @@
+  sub x0, x3, x11
+  ret
+.LTcmallocSlab_PushBatch_FixedShift_region2:
+  mov x0, #0
+  ret
+.LTcmallocSlab_PushBatch_FixedShift_region3:
+  .cfi_endproc
+ENCODE_SIZE(TcmallocSlab_PushBatch_FixedShift)
+DEFINE_UPSTREAM_CS(TcmallocSlab_PushBatch_FixedShift)
+
+/* size_t TcmallocSlab_PopBatch_FixedShift(
+ *     void *ptr (x0),
+ *     size_t cl (w1),
+ *     void** batch (x2),
+ *     size_t len (w3) {
+ *   uint64_t r8 = __rseq_abi.cpu_id
+ *   uint64_t* r8 = CpuMemoryStart(rdi, r8)
+ *   Header* hdr = GetHeader(rdi, rax, cl)
+ *   uint64_t r9 = hdr->current
+ *   uint64_t r10 = hdr->begin
+ *   if (r9 <= r10) return 0
+ *   r11 = min(rcx, r9 - r10)
+ *   rax = 0
+ * loop:
+ *   r9--
+ *   r10 = *(r8 + r9 * 8)
+ *   batch[rax] = r10
+ *   rax++
+ *   if (rax != r11) goto loop
+ *   hdr->current = r9
+ *   return rax
+ * }
+ */
+  .p2align 6 /* aligns to 2^6 with NOP filling */
+  .globl TcmallocSlab_PopBatch_FixedShift
+  .type  TcmallocSlab_PopBatch_FixedShift, @function
+TcmallocSlab_PopBatch_FixedShift:
+  .cfi_startproc
+.LTcmallocSlab_PopBatch_FixedShift_region0:
+  START_RSEQ(TcmallocSlab_PopBatch_FixedShift)
+  FETCH_CPU(w8)
+  lsl x8, x8, #PERCPU_TCMALLOC_FIXED_SLAB_SHIFT /* multiply cpu by 256k */
+  add x8, x0, x8
+  add x4, x8, x1, LSL #3
+  ldrh w9, [x4] /* current */
+  ldrh w10, [x4, #4] /* begin */
+  cmp w10, w9
+  bhs .LTcmallocSlab_PopBatch_FixedShift_region2
+  sub w11, w9, w10 /* w11 = available items */
+  cmp w3, w11
+  csel w11, w3, w11, ls /* r11 = min(len, available items) */
+  mov x12, #0
+.LTcmallocSlab_PopBatch_FixedShift_loop:
+  sub x9, x9, #1
+  ldr x10, [x8, x9, LSL #3]
+  str x10, [x2, x12, LSL #3]
+  add x12, x12, 1
+  cmp x12, x11
+  bne .LTcmallocSlab_PopBatch_FixedShift_loop
+  strh w9, [x4]
+.LTcmallocSlab_PopBatch_FixedShift_region1:
+.LTcmallocSlab_PopBatch_FixedShift_commit:
+  mov x0, x12
+  ret
+.LTcmallocSlab_PopBatch_FixedShift_region2:
+  mov x0, #0
+  ret
+.LTcmallocSlab_PopBatch_FixedShift_region3:
+  .cfi_endproc
+ENCODE_SIZE(TcmallocSlab_PopBatch_FixedShift)
+DEFINE_UPSTREAM_CS(TcmallocSlab_PopBatch_FixedShift)
+
+  .globl TcmallocSlab_Push
+  .type TcmallocSlab_Push, @function
+TcmallocSlab_Push:
+.LTcmallocSlab_Push_entry:
+  .cfi_startproc
+    // Arguments use:
+    //  *  x0: (Argument: Slabs*) cpu_0_slab_ptr
+    //  *  x1: (Argument: uintptr_t) cl
+    //  *  x2: (Argument: uintptr_t) p
+    //  *  w3: (Argument: size_t) shift
+    //  *  x4: (Argument: uintptr_t) f
+    // Return value: current CPU
+    // Available x5-x15
+
+    START_RSEQ(TcmallocSlab_Push)
+    FETCH_CPU(w8)
+    lsl x9, x8, x3
+    add x9, x0, x9
+    add x10, x9, x1, LSL #3
+    ldrh w12, [x10] /* current */
+    ldrh w11, [x10, #6] /* end */
+    cmp w11, w12
+    ble	.LTcmallocSlab_Push_no_capacity
+    str x2, [x9, x12, LSL #3]
+    add w12, w12, #1
+    strh w12, [x10]
+.LTcmallocSlab_Push_commit:
+    mov x0, x8
+    ret
+.LTcmallocSlab_Push_no_capacity:
+    mov x0, x8
```

Comment:
You could hoist this instruction above the `ble` to save code size.

Reply:
Same issue again with overwriting x0 in critical section.